acvm/src/lib.rs

```diff
@@ -150,6 +159,7 @@ pub trait ProofSystemCompiler {
         circuit: &Circuit,
         witness_values: WitnessMap,
         proving_key: &[u8],
+        is_recursive: bool,
```

Comment:
Why is this needed and can it not be deduced from the circuit?

Reply:
Unfortunately it cannot really be deduced from the circuit. An inner circuit may require a different prover in order to be verified recursively (in the case of bberg the inner circuit prover must be based upon Pedersen-Blake3 rather than Keccak). The internal circuit is able to just be a normal circuit that looks no different from any other Noir circuit

Reply:
what does is_recursive denote/mean?

Reply:
It means we want to generate a circuit which will be recursively verified in another circuit. But how it is being used right now is essentially to determine the hash type of the prover for an inner circuit. I do agree though this is not general enough and I will consider if there is a better way to abstract this functionality. What are your thoughts on a `hash_type` flag that determines the fiat shamir hash used by the prover?

Reply:
So the inner proof that is generated may use a different algorithm than a normal proof that can be verified by the Solidity verifier. The inner proof in this case may use a different hash function for Fiat-Shamir as that hash is more snark friendly and more optimized for recursive verification. 

This is information that would be passed to the proof system, however, there is no way to indicate from the circuit alone that we want to use a different hash type for the prover. I think if the `is_recursive` flag is too specific (and I agree is a bit of a leakage from bberg) we will need some other flag. This flag can either be a `hash_type` flag as proposed above or a custom data index of some sort. 

Perhaps a custom_data index of type [u8] would be simplest and then we just leave it for whichever backend is implementing these methods to indicate what custom data means. I imagine we would then want the backend to expose some sort of custom_data API as nargo is going to have to integrate these methods

Reply:
Its mainly because when you create proofs that you want to be natively verfied, you use perhaps sha256 for your hash function in the proof system. But when you want to verify the proof in a snark, you want the hash function being used to be snark friendly for the fiat-shamir/transcript. is_recursive will use a snark-friendly hash function when true and otherwise it will use the regular one, which for bb is keccak, but it can be anything for a proving system

Reply:
I added some comments to explain `is_recursive`